server/src/main/java/org/elasticsearch/search/vectors/ESDiversifyingChildrenKnnVectorQuery.java

```diff
@@ -0,0 +1,251 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.search.vectors;
+
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.search.DocIdSetIterator;
+import org.apache.lucene.search.Explanation;
+import org.apache.lucene.search.IndexSearcher;
+import org.apache.lucene.search.MatchNoDocsQuery;
+import org.apache.lucene.search.Query;
+import org.apache.lucene.search.QueryVisitor;
+import org.apache.lucene.search.ScoreMode;
+import org.apache.lucene.search.Scorer;
+import org.apache.lucene.search.Weight;
+import org.apache.lucene.search.join.BitSetProducer;
+import org.apache.lucene.search.join.DiversifyingChildrenByteKnnVectorQuery;
+import org.apache.lucene.search.join.DiversifyingChildrenFloatKnnVectorQuery;
+import org.apache.lucene.util.BitSet;
+
+import java.io.IOException;
+import java.util.Objects;
+
+import static org.apache.lucene.search.DocIdSetIterator.NO_MORE_DOCS;
+
+/**
+ * This query is used to score vector child documents given the results of a {@link DiversifyingChildrenByteKnnVectorQuery}
+ * or {@link DiversifyingChildrenFloatKnnVectorQuery}.
+ */
+public class ESDiversifyingChildrenKnnVectorQuery extends Query {
```

Comment:
I don't understand why this query is needed? Is it to avoid scoring the children returned by the query phase? 
I don't think we should use any of the child result of the query phase to compute the inner hits. Ideally we can just have a nested query builder that uses a brute force knn query to score all children. We don't need to bother with the DiversifyingChildren... queries in the fetch phase.

Reply:
@jimczi, for knn query, there is no way I could find to interject information without leaking the abstraction up to the nested query that it somewhere contains kNN results. We don't rewrite to a `KnnScoreAndDoc` query when using knn query. We just use what Lucene gives us, so we need to wrap the query lower down which is what I did here.

Reply:
Ouch, thanks I forgot this part. I am conflicted because this solution involves running the approximate query twice. The first time during the query phase and then another time during the inner hits fetch sub-phase. This is more an issue at the Lucene query level since we rely on the rewrite to do the heavy lifting and we have no way to determine the context where the rewrite is executed.

Reply:
> I am conflicted because this solution involves running the approximate query twice. 

Ah, so `nested` will run the query again, completely from scratch when gathering the inner_hits? That is indeed frustrating. 

I will think a bit more on this. Maybe there is something we can do in Elasticsearch.

Reply:
OK, looking at the top-level call extract is done here: 
https://github.com/elastic/elasticsearch/blob/f33122a191c664622295adf6c38ee37973b81f6e/server/src/main/java/org/elasticsearch/search/SearchService.java#L1238

This has access to a `SearchExecutionContext` and such. Which indicates to me we *could* have an additional thing that `NestedQueryBuilder` could call like `rewriteForInnerHits`. Multi-queries would have to push this down (but I think it would be easy enough), but then `KnnVectorQueryBuilder` could satisfy this interface and return a separate query builder that is a BruteForceQueryBuilder like you have. 

What do you think?

Reply:
OK, I put this attempt in the latest commit @jimczi . What do you think?

One of the downsides is that this seems to completely break "total_hits". Even when pre-filtering with a match that only hits one document, when calculating inner_hits, the top-level total hits is overwritten to the number of all parent docs in the index. 

It seems weird to me that fetching particular parent doc IDs would update the top level total hit count at all.

Reply:
Ah, I reverted something I shouldn't have. I will address and push again. I think this will work.

Reply:
Thanks @benwtrent I like the idea. Another option would be to add a new `QueryRewriteContext` so that we don't need to implement the new functions on all multi-queries. The advantage would be that it would also work on custom queries added via plugin without requiring any change?
Just for completeness we could also consider having a knn queries that executes at the scorerSupplier level (we know the leadCost there so we could pick between exact and approximate) but that would be a bigger change and would also change the expectation (each segment would return topN results).

Reply:
> Another option would be to add a new QueryRewriteContext so that we don't need to implement the new functions on all multi-queries. 

I thought of that option as well. I will see if it will work. I am not sure at what point the fetch will rewrite these queries.

Reply:
@jimczi added a "InnerHitsRewriteContext" and rewrite with it in the SearchService and call extractInnerHits on the rewritten query.